### PR TITLE
Add Tehran prayer times

### PR DESCRIPTION
**Pull request by basememara**
Involved adding custom Maghrib angle as Fajr and Isha options. This is a continuation of previous PR https://github.com/batoulapps/Adhan/pull/69 from the old branch.

### CODE REVIEW

File: Sources/PrayerTimes.swift [hunk 1]
```diff
@@ -137,6 +137,16 @@ public struct PrayerTimes {
...
+        
+        // Maghrib calculation with check against safe value
+        if let maghribAngle = calculationParameters.maghribAngle,
+            let maghribComponents = solarTime.timeForSolarAngle(Angle(-maghribAngle), afterTransit: true),
+            let maghribDate = cal.date(from: maghribComponents),
+            // maghrib safe if falls between asr and maghrib
+            (tempAsr?.compare(maghribDate) == .orderedAscending || tempAsr == nil),
+            (tempIsha?.compare(maghribDate) == .orderedDescending || tempIsha == nil) {
+                tempMaghrib = maghribDate
+        }
```
Comment (basememara) [inline Sources/PrayerTimes.swift]: Good call on getting rid of `tempSunset` and refactoring this.

It's hard to test the edge cases, so here's the original code to compare next to:

```
        // Maghrib calculation with check against safe value
        if let maghribAngle = calculationParameters.maghribAngle {
            if let maghribComponents = solarTime.timeForSolarAngle(Angle(-maghribAngle), afterTransit: true) {
                tempMaghrib = cal.date(from: maghribComponents)
            }
            
            if let tempAsr = tempAsr, let tempIsha = tempIsha {
                // maghrib safe if falls between asr and maghrib
                if tempMaghrib == nil || tempMaghrib?.compare(tempAsr) == .orderedAscending || tempMaghrib?.compare(tempIsha) == .orderedDescending {
                    tempMaghrib = tempSunset
                }
            } else {
                // fallback to regular sunset
                tempMaghrib = tempSunset
            }
        } else {
            tempMaghrib = tempSunset
        }
```

File: Sources/Models/Madhab.swift [hunk 2]
```diff
@@ -29,13 +29,16 @@ import Foundation
 public enum Madhab {
     case shafi
     case hanafi
+    case jafari
```
[the remark this answers is not in the document]
Comment (basememara) [reply Sources/Models/Madhab.swift]: My thought was just clarity for the user and possible future optimizations to this madhab route. Might be weird for the user to pick these times with a different Madhab they are intending. Thoughts from here?

**Comment by basememara** [comment]
Jazak'Allah for the feedback. Will handle the obj-c route in a bit insha'Allah...

File: Sources/PrayerTimes.swift (hunk 1 above)
Comment (basememara) [reply Sources/PrayerTimes.swift]: Does a negative Maghrib angle make sense, or do I have to still check that it's greater than zero?

File: Sources/Models/Madhab.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (basememara) [reply Sources/Models/Madhab.swift]: You know on second thought, there's no option for Maliki or Hanbali, so it's the same. Ill remove it, let me know if my previous comment convinced you and I'll add it back :)

[the remark this answers is not in the document]
Comment (basememara) [reply Sources/Models/Madhab.swift]: I think at one point you had `standard` instead of `shafi`, which I think makes more sense than users thinking "my madhab isn't supported".

File: Sources/PrayerTimes.swift (hunk 1 above)
Comment (z3bi) [reply Sources/PrayerTimes.swift]: In reality we are forcing it to negative `Angle(-maghribAngle)`, because you're looking for an angle below the horizon, if they provided a negative value and then we forced it to positive it would be above the horizon and produce a time earlier than sunset. 

If we can better define what the true values are, we can have a better sanity check.

File: Sources/Models/Madhab.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (z3bi) [reply Sources/Models/Madhab.swift]: To change the case right now would be a breaking change, I think it most makes sense to leave a comment above the enum case stating that it is also valid for Maliki, Hanbali, and Jafari

File: Sources/PrayerTimes.swift (hunk 1 above)
Comment (z3bi) [reply Sources/PrayerTimes.swift]: If we can find better information about what this is supposed to do that would help, it would seem Jafari maghrib probably requires it to be post sunset, so maybe a better comparison is between sunset and Isha.

Also in Swift `Date` now conforms to `Comparable` so we don't need to use the `compare()` function, we can just use the `<` and `>` operators.

File: Sources/AdhanObjc.swift [hunk 3]
```diff
@@ -127,30 +127,33 @@ import Foundation
 @objcMembers open class BACalculationParameters: NSObject {
     open var method: BACalculationMethod = .other
     open var fajrAngle: Double
+    open var maghribAngle: Double?
```
Comment (z3bi) [inline Sources/AdhanObjc.swift]: let's add a test in the objc tests please. Also We're still missing a case in `BACalculationMethod`

File: Sources/PrayerTimes.swift (hunk 1 above)
Comment (basememara) [reply Sources/PrayerTimes.swift]: I'll find out more answers about this. For comparable, just didn't want to deal with optionals since its not comparable.

Comment (z3bi) [reply Sources/PrayerTimes.swift]: Yeah that makes sense

Comment (basememara) [reply Sources/PrayerTimes.swift]: Testing Maghrib after sunset is a safer test than after Asr, so I made the change. This is the details of the formula: http://praytimes.org/calculation

File: Sources/AdhanObjc.swift (hunk 3 above)
Comment (basememara) [reply Sources/AdhanObjc.swift]: Done

**Comment by z3bi** [comment]
So one thing is that `BACalculationParameters` in the Objective-C interface can't have an optional, as that is not compatible with Objective-C. What I would suggest is making that a double with a default value of 0, but then check for 0 when constructing the real params. Something like:
```
params.maghribAngle = self.maghribAngle == 0 ? nil : self.maghribAngle
```

**Comment by basememara** [comment]
Ok I tried to fix the Obj-C optional, but I don't know how to check if this worked. Since the unit tests were passing already for Obj-C. Also wondering why `BAPrayerTimes` works and it has optionals? Sorry I'm horrible at Obj-C ;D

**Comment by basememara** [comment]
..oh you probably mean the function parameters can't be optional

**Comment by z3bi** [comment]
So in objective-c objects can be optional, because it's possible for them to be null. However primitives, such as Double, can't be null in objective-c, so that's the issue.

**Comment by z3bi** [comment]
You make a good point that the unit tests were passing already for Obj-C, that's because we were only initializing params via a calculation method. It would make sense to add tests that created params manually, probably both for Swift and Objective-C. Is this something you would feel comfortable adding?

**Comment by basememara** [comment]
It wouldn’t be safe for me to touch the Obj-C, I jumped straight from web to Swift so I skipped the Obj-C era. Also maybe we could do as a separate PR and see if there’s anything left with this feature (unless you want to jump into the branch and add them if it’s quick)?

**Comment by z3bi** [comment]
That's a fair point. I think to finish up this PR we need to add the new methods to the README and update it to include the maghrib angle parameter.

**Comment by basememara** [comment]
Ah good catch, let me know if you want further changes with the readme or anything.

File: README.md [hunk 4]
```diff
@@ -97,6 +97,7 @@ params.adjustments.fajr = 2
 | --------- | ----------- |
 | method    | Member of CalculationMethod enum |
 | fajrAngle | Angle of the sun used to calculate Fajr |
+| maghribAngle | Angle of the sun used to calculate Maghrib |
```
Comment (z3bi) [inline README.md]: Let's clarify with: `Angle of the sun used to calculate Maghrib for some Calculation Methods`

File: README.md [hunk 5]
```diff
@@ -116,14 +117,16 @@ params.adjustments.fajr = 2
 | kuwait | Method used by the country of Kuwait. Fajr angle: 18, Isha angle: 17.5 |
 | moonsightingCommittee | Moonsighting Committee. Fajr angle: 18, Isha angle: 18. Also uses seasonal adjustment values. |
 | singapore | Method used by Singapore. Fajr angle: 20, Isha angle: 18. |
+| tehran | Institute of Geophysics, University of Tehran. Fajr angle: 17.7, Maghrib angle: 4.5, Isha angle: 14. |
+| shia | Shia Ithna-Ashari, Leva Institute, Qum. Fajr angle: 16, Maghrib angle: 4, Isha angle: 14. |
```
Comment (z3bi) [inline README.md]: The tehran calculation method makes perfect sense because it's specific angle that is valid for a specific location. I have reservations about including a method with extremely different values (isha angle of 14 is a very different value), without it being for a specific region. I think right now it would be best if we only added the tehran method.
